### DISCUSSION TITLE
ClassDataSource - Is it possible to have two fixtures at different scopes on the same class?

### DISCUSSION THREAD

**Discussion by richardoSGSI** [thomhurst/TUnit#991] [Q&A] [answered]
For example,

```
namespace MyTests
{
  [ClassDataSource<MyFixture>(Shared = SharedType.Globally)]
  [ClassDataSource<AnotherFixture>(Shared = SharedType.ForClass)]
  public class MyTestClass
  {
    ...
  }
}
```

**Comment by thomhurst** [comment, marked as answer]
As part of the same instance (so injected at the same time)? Or two separate invocations, which would end up as different test cases?

For the former, no. For the latter, yes, and that's what the above would do.

If you did want the former, you could use property injection instead.

```csharp
namespace MyTests
{
  public class MyTestClass
  {
    [ClassDataSource<MyFixture>(Shared = SharedType.Globally)]
    public required MyFixture Fixture1 { get; init; }
  
    [ClassDataSource<AnotherFixture>(Shared = SharedType.ForClass)]
    public required AnotherFixture Fixture2 { get; init; }
  }
}
```

**Reply by richardoSGSI** [reply]
@thomhurst 

is there a difference between adding the attribute on the class vs a property? am I able to do things with one that I can't do with the other?

**Reply by thomhurst** [reply]
Property injection gives you the ability to mix scopes, but apart from that, there's not really any difference.

It can be useful where you want to define a base class but don't want to have to repeat your constructor in all your derived classes.
But some people will always prefer constructor injection - So you have the choice of what to use 😄
